### PR TITLE
added birth death migration

### PR DESCRIPTION
**Pull request by jinpz**

### CODE REVIEW

File: scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration.tmcf [hunk 1]
```diff
@@ -0,0 +1,49 @@
+
+Node: E:EurostatNUTS3_BirthDeathMigration->E0
+typeOf: dcs:StatVarObservation
+variableMeasured: dcs:Count_MortalityEvent
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration.tmcf:4]: Can you reuse https://datacommons.org/browser/Count_Death

File: scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration.tmcf [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+observationAbout: C:EurostatNUTS3_BirthDeathMigration->geo
+observationDate: C:EurostatNUTS3_BirthDeathMigration->time
+value: C:EurostatNUTS3_BirthDeathMigration->Count_MortalityEvent
+
+Node: E:EurostatNUTS3_BirthDeathMigration->E1
+typeOf: dcs:StatVarObservation
+variableMeasured: dcs:Count_BirthEvent_AsAFractionOfCount_Person
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration.tmcf:11]: `s/dcs:Count_BirthEvent_AsAFractionOfCount_Person/dcs:Count_BirthEvent_AsAFractionOf_Count_Person/g` (more readable, decided to use this syntax going forward

File: scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration.tmcf (hunk 1 above)
Comment (tjann) [reply scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration.tmcf:4]: Similar applies below

File: scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration_statvar.mcf [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+typeOf: dcs:StatisticalVariable
+populationType: dcs:BirthEvent
+measuredProperty: dcs:count
+statType: dcs:measuredValue
+
+Node: dcid:Count_MortalityEvent
+typeOf: dcs:StatisticalVariable
+populationType: schema:MortalityEvent
+measuredProperty: dcs:count
+statType: dcs:measuredValue
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration_statvar.mcf:17]: Could you remove these variables since they already exist? We recently solved the inconvenience of having to redefine existing variables, so you actually don't need to do this at all anymore.

File: scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration_statvar.mcf [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+populationType: dcs:MortalityEvent
+measuredProperty: dcs:count
+statType: dcs:measuredValue
+measurementDenominator: dcs:Count_Person
+
+Node: dcid:GrowthRate_Count_Person
+typeOf: dcs:StatisticalVariable
+populationType: schema:Person
+measuredProperty: dcs:count
+statType: dcs:growthRate
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration_statvar.mcf:43]: This also exists

File: scripts/eurostat/birth_death_migration/Readme.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# Importing Eurostat Demographic balance and crude rates into Data Commons
+
+Author: jinpz
+
+## Table of Contents
+
+1. [About the Dataset](#about-the-dataset)
+1. [Import Procedure](#Import-Procedure)
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/Readme.md:8]: This link is broken, maybe bc of caps?

File: scripts/eurostat/birth_death_migration/Readme.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+```
+python3 import_data.py
+```
+
+### Testing Procedure
+
+To test import procedure, run the following command:
+```
+python3 test_import.py
+```
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/Readme.md:55]: Very nice readme. Thanks.

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    data frame.
+    """
+    raw_df = pd.read_table(DATA_LINK)
+    return raw_df
+
+
+def preprocess_data(raw_df):
+    """Preprocesses instance raw_df and puts it into long format."""
+    if raw_df is None:
+        raise ValueError("Uninitialized value of raw data frame. Please "
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:29]: This error message seems to assume context of the entire script. Instead, you might raise a ValueError simply stating that input `raw_df` should not be None?

Comment (tjann) [reply scripts/eurostat/birth_death_migration/import_data.py:29]: Could even just be an assert like below.

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+def download_data():
+    """Downloads raw data from Eurostat website and stores it in instance
+    data frame.
+    """
+    raw_df = pd.read_table(DATA_LINK)
+    return raw_df
+
+
+def preprocess_data(raw_df):
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:26]: Since there is some amount of specific expectations in this func, can add a longer docstring about what preprocessing you are doing, and try to help differentiate preprocess from clean functions.

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        "([0-9:])$", lambda m: m.group(0) + ' ')
+    statistical_variable = None
+    # In case this returns only one element in the list.
+    first_column_list = preprocessed_df.columns[0].rsplit(sep=",", maxsplit=1)
+    if len(first_column_list) == 2:
+        statistical_variable, geo = first_column_list
+    else:
+        geo = first_column_list[0]
+    geo = geo.replace(r'\time', '')
+    assert geo == "geo", "Expected the column header to end with 'geo'."
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:56]: More readable: column header should end with 'geo'

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                         "before clean_data.")
+    # drop unused columns
+    clean_df = preprocessed_df[preprocessed_df.columns[:len(
+        preprocessed_df.columns
+    ) // 2 + 1]]  # number of columns should be 2 + 2X, we want the first 2 + X
+    # columns
+    # replace colon with NaN.
+    clean_df = clean_df.replace(':', '')
+    clean_df.to_csv(output_path, index=False)
+    clean_df = pd.read_csv(output_path)
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:105]: What is happening here? you do another to_csv later too.

File: scripts/eurostat/birth_death_migration/test/test1/data.tsv [hunk 11]
```diff
@@ -0,0 +1,300 @@
+indic_de,geo\time	2019 	2018 	2017 	2016 	2015 	2014 	2013 	2012 	2011 	2010 	2009 	2008 	2007 	2006 	2005 	2004 	2003 	2002 	2001 	2000
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/test/test1/data.tsv:1]: Do we need 300 rows? For testing, it should be doable by hand. Please select a few but good cases (look for edge cases or variety).

File: scripts/eurostat/birth_death_migration/test_import.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+    def test1(self):
+        """Simple unit test on melting csv content"""
+        self._test_preprocess_output('./test/test1')
+
+    def test2(self):
+        """Simple integration test on output csv content"""
+        self._test_csv_output('./test/test2')
+
+    def _test_preprocess_output(self, dir_path: str):
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/test_import.py:32]: I think generally prefer helper functions to be defined before.

File: scripts/eurostat/birth_death_migration/test_import.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        Returns:
+            str: expected output == actual output.
+        """
+
+        input_path = os.path.join(dir_path, "data.tsv")
+        output_path = os.path.join(dir_path, "output.csv")
+        expected_path = os.path.join(dir_path, "expected.csv")
+
+        if not os.path.exists(input_path):
+            self.fail(input_path + " doesn't exist!")
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/test_import.py:48]: Cool. I learned something new today (self.fail)

File: scripts/eurostat/birth_death_migration/test_import.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        if not os.path.exists(expected_path):
+            self.fail(expected_path + " doesn't exist!")
+
+        # Generate the output csv file.
+        input_df = pd.read_csv(input_path, sep='\s*\t\s*', engine='python')
+        preprocess_data(input_df).to_csv(output_path, index=False)
+        # Get the content from the csv file.
+        actual_f = open(output_path, 'r+')
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/test_import.py:56]: Prefer using "with" keyword to handle automatic closing: https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files

File: scripts/eurostat/birth_death_migration/test_import.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+        # Get the content of the expected output.
+        expected_f = open(expected_path, 'r+')
+        expected = expected_f.read()
+        expected_f.close()
+
+        self.assertEqual(actual, expected)
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/test_import.py:104]: Nicely done. Just please cut down the 300 line input csv.

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    clean_df = clean_df[original_names]
+    clean_df[['GBIRTHRT', 'GDEATHRT',
+              'GROWRT']] /= 1000  # apply scaling factor of 1000
+    clean_df.columns = new_names
+    clean_df.to_csv(output_path, index=False)
+
+
+if __name__ == '__main__':
+    DATA_LINK = "https://ec.europa.eu/eurostat/estat-navtree-portlet-prod/BulkDownloadListing?file=data/demo_r_gind3.tsv.gz"
+    FINAL_CSV_NAME = 'demo_r_gind3_final_stage.csv'
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:128]: Please move constants to the top of the file (under the imports). Also, add a _ if the constant is not used outside this file. You may revisit https://google.github.io/styleguide/pyguide.html when you have some time.

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+                         "preprocess_data.")
+    assert len(raw_df.columns) > 1, "Data must have at least two columns."
+    assert raw_df.columns.values[0].endswith(
+        '\\time'), "Expected the first column header to end with '\\time'."
+
+    # Convert from one-year per column (wide format) to one row per
+    # data point (long format).
+    preprocessed_df = raw_df.melt(id_vars=[raw_df.columns[0]])
+    # Rename the variable column.
+    preprocessed_df.columns.values[1] = 'time'
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:40]: Is it safer to do a column rename() mapping {original_name: 'time'} ? As opposed to relying on index.

Another thing you should consider is asserting the expected column names before doing preprocessing. Sometimes sources may change their columns.

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+
+    # Split the index column into two.
+    # Ex: 'unit,sex,age,geo\time' -> 'unit,sex,age' and 'geo'.
+    # '\time' labels the other columns so it is confusing.
+    # Replace value column with extra space.
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:45]: Please also explain why this is needed.

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    preprocessed_df.value = preprocessed_df.value.str.replace(
+        "([0-9:])$", lambda m: m.group(0) + ' ')
+    statistical_variable = None
+    # In case this returns only one element in the list.
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:49]: Maybe just explain that sometimes there is an extra `,geo` after variable for places like EU2017 where the year determines what members are in the EU. That's the issue here, right?

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+    first_column_list = preprocessed_df.columns[0].rsplit(sep=",", maxsplit=1)
+    if len(first_column_list) == 2:
+        statistical_variable, geo = first_column_list
+    else:
+        geo = first_column_list[0]
+    geo = geo.replace(r'\time', '')
+    assert geo == "geo", "Expected the column header to end with 'geo'."
+
+    if statistical_variable:
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:58]: Are there any cases where there is no statvar?

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+        split_df = preprocessed_df[preprocessed_df.columns[0]].str.rsplit(
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:59]: For readability and safety, please use column names instead of indices whenever possible

File: scripts/eurostat/birth_death_migration/import_data.py [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    preprocessed_df[preprocessed_df.columns[2:]] = preprocessed_df[
+        preprocessed_df.columns[2:]].replace("([0-9:])$",
+                                             lambda m: m.group(0) + ' ')
+    # replace comma in column names with a vertical bar so that we can save it in csv later
+    new_column_names = []
+    for column_name in preprocessed_df.columns:
+        new_column_names.append(column_name.replace(',', '|'))
+    preprocessed_df.columns = new_column_names
+    for column_name in preprocessed_df.columns[2:]:
+        preprocessed_df['{0}'.format(column_name)], preprocessed_df[
```
Comment (tjann) [inline scripts/eurostat/birth_death_migration/import_data.py:85]: Can you explain what's happening here?

File: scripts/eurostat/birth_death_migration/Readme.md (hunk 5 above)
Comment (jinpz) [reply scripts/eurostat/birth_death_migration/Readme.md:8]: just tested it on my fork, it seems to be okay. For consistency, I replaced the caps. Where would it break?

File: scripts/eurostat/birth_death_migration/import_data.py (hunk 7 above)
Comment (jinpz) [reply scripts/eurostat/birth_death_migration/import_data.py:29]: decided to remove it since it is not very meaningful

File: scripts/eurostat/birth_death_migration/import_data.py (hunk 20 above)
Comment (jinpz) [reply scripts/eurostat/birth_death_migration/import_data.py:58]: For other eurostat dataset, yes. There was one case where the table does not have a statvar. So it's just a flat 2d table

File: scripts/eurostat/birth_death_migration/import_data.py (hunk 16 above)
Comment (jinpz) [reply scripts/eurostat/birth_death_migration/import_data.py:128]: decided to refactor these to be normal variable since it's only used once.

File: scripts/eurostat/birth_death_migration/import_data.py (hunk 19 above)
Comment (jinpz) [reply scripts/eurostat/birth_death_migration/import_data.py:49]: Thanks for pointing this out. This was for adding a whitespace after each value, so that when we split value and notes, the note column will get a white space (otherwise error will be raised). But I found out I have the same line below so this is redundant (now deleted)

File: scripts/eurostat/birth_death_migration/import_data.py (hunk 21 above)
Comment (jinpz) [reply scripts/eurostat/birth_death_migration/import_data.py:59]: Agreed. It's just this preprocess_data function is kept a bit general so that even if they change the column names, as long as the order is maintained, the script can still run. That is why I didn't hard code the column name here (which should be 'indic_de,geo\time')

File: scripts/eurostat/birth_death_migration/import_data.py (hunk 10 above)
Comment (jinpz) [reply scripts/eurostat/birth_death_migration/import_data.py:105]: Thanks for the catch. I must have accidentally committed this when doing the testing

File: scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration.tmcf (hunk 2 above)
Comment (tjann) [reply scripts/eurostat/birth_death_migration/EurostatNUTS3_BirthDeathMigration.tmcf:11]: You might have missed this one: AsAFractionOfCount_Person should be AsAFractionOf_Count_Person

File: scripts/eurostat/birth_death_migration/import_data.py (hunk 21 above)
Comment (tjann) [reply scripts/eurostat/birth_death_migration/import_data.py:91]: We would rather fail than have the wrong value, which is why we prefer names. Does that make sense?

File: scripts/eurostat/birth_death_migration/import_data.py (hunk 22 above)
Comment (tjann) [reply scripts/eurostat/birth_death_migration/import_data.py:114]: statistical value: since statvar is a concept, we try to use it precisely.

File: scripts/eurostat/birth_death_migration/import_data.py (hunk 20 above)
Comment (tjann) [reply scripts/eurostat/birth_death_migration/import_data.py:90]: I see.

**Review by tjann — APPROVED** [review]
Thanks, Jin!
